Make MessageEvent and ExtendableMessageEvent hold a SecurityOrigin

annevk opened the pull request:
#### 3406f735d02b02f588b0c954e9ccce27b23221df
<pre>
Make MessageEvent and ExtendableMessageEvent hold a SecurityOrigin
<a href="https://bugs.webkit.org/show_bug.cgi?id=303629">https://bugs.webkit.org/show_bug.cgi?id=303629</a>
<a href="https://rdar.apple.com/165916784">rdar://165916784</a>

Reviewed by Alex Christensen.

This is meant to be a strict refactoring driven by the needs of the
Origin API (bug 302785) which needs to be able to derive consistent
opaque origins from message events.

As such this should be covered by existing tests.

In MessageEvent.cpp we also tidy up an existing WTF::switchOn.

Canonical link: <a href="https://commits.webkit.org/304156@main">https://commits.webkit.org/304156@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4d05be537cc83083e30994b89e4a20653f425243

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134725 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7169 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45991 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142245 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86650 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4c8b660f-ea5f-4fc3-8925-f0b19e7ba999) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136595 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7779 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7021 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102966 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/79f8523f-e471-4698-afb4-8afe1a59485d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137672 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5473 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120770 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83776 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2a9cbc09-28ec-4f08-9316-8dd8ec0d6495) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5319 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2928 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2834 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114548 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38917 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144940 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6845 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39498 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6916 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5739 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111660 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28325 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5154 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117050 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60736 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6894 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35235 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6695 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70474 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6931 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6804 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/websockets/WebSocket.h:
```diff
@@ -137,6 +137,7 @@ class WebSocket final : public RefCounted<WebSocket>, public EventTarget, public
 
     State m_state { CONNECTING };
     URL m_url;
+    RefPtr<SecurityOrigin> m_origin;
```
We could lazily create it, either in connect or when we need to dispatch an event and mark m_origin as `const RefPtr`.

youennf commented on Source/WebCore/dom/MessageEvent.h:
```diff
@@ -90,10 +91,11 @@ class MessageEvent final : public Event {
 
     DataType m_data WTF_GUARDED_BY_LOCK(m_concurrentDataAccessLock);
-    String m_origin;
+    String m_misleadingOrigin;
```
Is the term misleading origin in a spec?
Should it be m_overridingOrigin instead?

annevk replied:
I went with overriding for now. Eventually I'd like this to have a name that indicates it's less trusted and also put that in the specification (which is yet to be updated). Maybe `m_userlandOrigin`?

youennf commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -1058,7 +1058,7 @@ void LocalDOMWindow::postMessageFromRemoteFrame(JSC::JSGlobalObject& lexicalGlob
         targetOrigin = targetOriginData->securityOrigin();
 
-    processPostMessage(lexicalGlobalObject, sourceOrigin, message, WTFMove(source), WTFMove(targetOrigin));
+    processPostMessage(lexicalGlobalObject, SecurityOrigin::create(sourceOrigin.isolatedCopy()), message, WTFMove(source), WTFMove(targetOrigin));
```
Why are we isolatedCopying sourceOrigin?
Can we just do `sourceOrigin.securityOrigin()` instead?

youennf commented on Source/WebCore/workers/service/ExtendableMessageEvent.cpp:
```diff
@@ -65,23 +66,25 @@ auto ExtendableMessageEvent::create(JSC::JSGlobalObject& globalObject, Vector<Re
     : ExtendableEvent(EventInterfaceType::ExtendableMessageEvent, type, init, isTrusted)
-    , m_origin(init.origin)
+    , m_misleadingOrigin(init.origin)
+    , m_origin(nullptr)
```
```suggestion
```

youennf commented on Source/WebCore/workers/service/ExtendableMessageEvent.cpp:
```diff
@@ -65,23 +66,25 @@ auto ExtendableMessageEvent::create(JSC::JSGlobalObject& globalObject, Vector<Re
-ExtendableMessageEvent::ExtendableMessageEvent(const AtomString& type, const String& origin, const String& lastEventId, std::optional<ExtendableMessageEventSource>&& source, Vector<Ref<MessagePort>>&& ports)
+ExtendableMessageEvent::ExtendableMessageEvent(const AtomString& type, Ref<SecurityOrigin>&& origin, const String& lastEventId, std::optional<ExtendableMessageEventSource>&& source, Vector<Ref<MessagePort>>&& ports)
     : ExtendableEvent(EventInterfaceType::ExtendableMessageEvent, type, CanBubble::No, IsCancelable::No)
+    , m_misleadingOrigin(nullString())
```
I would tend to use the default constructor here.
Is there a difference between setting `m_misleadingOrigin` to nullString and using the default string constructor?

youennf commented on Source/WebCore/workers/service/ExtendableMessageEvent.h:
```diff
@@ -62,26 +63,27 @@ class ExtendableMessageEvent final : public ExtendableEvent {
     JSValueInWrappedObject m_data;
-    String m_origin;
+    String m_misleadingOrigin;
+    RefPtr<SecurityOrigin> m_origin;
```
can they both be `const`?

annevk replied:
I made `m_origin` `const`. You want `const String` as well?

youennf commented on Source/WebCore/dom/MessageEvent.h:
```diff
@@ -90,10 +91,11 @@ class MessageEvent final : public Event {
     DataType m_data WTF_GUARDED_BY_LOCK(m_concurrentDataAccessLock);
-    String m_origin;
+    String m_misleadingOrigin;
+    RefPtr<SecurityOrigin> m_origin;
```
Can they both be const?

annevk replied:
No, this has `initMessageEvent()` unfortunately.

youennf commented on Source/WebCore/dom/MessageEvent.cpp:
```diff
@@ -48,25 +49,27 @@ MessageEvent::MessageEvent()
     : Event(EventInterfaceType::MessageEvent, type, CanBubble::No, IsCancelable::No)
     , m_data(WTFMove(data))
-    , m_origin(origin)
+    , m_misleadingOrigin(nullString())
```
Is it needed?

youennf commented on Source/WebCore/dom/MessageEvent.cpp:
```diff
@@ -48,25 +49,27 @@ MessageEvent::MessageEvent()
     , m_data(JSValueTag { })
-    , m_origin(initializer.origin)
+    , m_misleadingOrigin(initializer.origin)
+    , m_origin(nullptr)
```
Not needed probably.

youennf commented on Source/WebCore/workers/service/SWClientConnection.cpp:
```diff
@@ -128,22 +129,22 @@ void SWClientConnection::startScriptFetchForServer(ServiceWorkerJobIdentifier jo
 {
     if (RefPtr container = context.ensureServiceWorkerContainer())
-        container->postMessage(WTFMove(message), WTFMove(sourceData), WTFMove(sourceOrigin));
+        container->postMessage(WTFMove(message), WTFMove(sourceData), SecurityOrigin::create(sourceOrigin.isolatedCopy()));
```
```suggestion
        container->postMessage(WTFMove(message), WTFMove(sourceData), sourceOrigin.securityOrigin());
```

youennf commented on Source/WebCore/workers/service/ServiceWorkerClient.cpp:
```diff
@@ -92,7 +92,7 @@ ExceptionOr<void> ServiceWorkerClient::postMessage(JSC::JSGlobalObject& globalOb
     Ref context = downcast<ServiceWorkerGlobalScope>(*scriptExecutionContext());
     auto sourceIdentifier = context->thread()->identifier();
-    callOnMainThread([message = WTFMove(message), destinationIdentifier = identifier(), sourceIdentifier, sourceOrigin = context->origin().isolatedCopy()] {
+    callOnMainThread([message = WTFMove(message), destinationIdentifier = identifier(), sourceIdentifier, sourceOrigin = context->protectedSecurityOrigin()->data()] {
```
We probably want to keep isolated copying context->origin().

annevk commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -1058,7 +1058,7 @@ void LocalDOMWindow::postMessageFromRemoteFrame(JSC::JSGlobalObject& lexicalGlob
         targetOrigin = targetOriginData->securityOrigin();
 
-    processPostMessage(lexicalGlobalObject, sourceOrigin, message, WTFMove(source), WTFMove(targetOrigin));
+    processPostMessage(lexicalGlobalObject, sourceOrigin.securityOrigin(), message, WTFMove(source), WTFMove(targetOrigin));
```
I guess I should WTFMove here, right?

achristensen07 replied:
no

annevk commented on Source/WebCore/workers/service/SWClientConnection.cpp:
```diff
@@ -128,22 +129,22 @@ void SWClientConnection::startScriptFetchForServer(ServiceWorkerJobIdentifier jo
 {
     if (RefPtr container = context.ensureServiceWorkerContainer())
-        container->postMessage(WTFMove(message), WTFMove(sourceData), WTFMove(sourceOrigin));
+        container->postMessage(WTFMove(message), WTFMove(sourceData), sourceOrigin.securityOrigin());
```
I guess I should WTFMove here, right?

achristensen07 replied:
no

annevk commented on Source/WebCore/workers/shared/SharedWorkerGlobalScope.cpp:
```diff
@@ -69,13 +69,13 @@ Ref<SharedWorkerThread> SharedWorkerGlobalScope::thread()
     ASSERT(ports.size() == 1);
     RefPtr port = ports[0].ptr();
-    auto event = MessageEvent::create(emptyString(), sourceOrigin, { }, port, WTFMove(ports));
+    auto event = MessageEvent::create(emptyString(), sourceOriginData.securityOrigin(), { }, port, WTFMove(ports));
```
I guess I should WTFMove here, right?

achristensen07 replied:
no

achristensen07 commented on Source/WebCore/workers/service/ExtendableMessageEvent.h:
```diff
@@ -62,26 +63,27 @@ class ExtendableMessageEvent final : public ExtendableEvent {
     JSValueInWrappedObject m_data;
-    String m_origin;
+    String m_overridingOrigin;
+    const RefPtr<SecurityOrigin> m_origin;
```
This could be a `Variant<String, Ref<SecurityOrigin>>` origin since we'll only ever have one or the other.  m_overridingOrigin isn't really overriding, it's just an alternative form for an origin and if we have it we never have a `Ref<SecurityOrigin>`

That might conflict with the next step to use the SecurityOrigin.  I'm not sure if this is actually a good idea.

annevk replied:
I really like the idea so I'm going to run with it for now.

annevk committed:
Make MessageEvent and ExtendableMessageEvent hold a SecurityOrigin
https://bugs.webkit.org/show_bug.cgi?id=303629
rdar://165916784

Reviewed by Alex Christensen.

This is meant to be a strict refactoring driven by the needs of the
Origin API (bug 302785) which needs to be able to derive consistent
opaque origins from message events.

As such this should be covered by existing tests.

In MessageEvent.cpp we also tidy up an existing WTF::switchOn.

Canonical link: https://commits.webkit.org/304156@main
